File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/sdk/screen/AppOpenMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.sdk.screen

import com.cleveradssolutions.plugin.flutter.CASFlutterContext
import com.cleveradssolutions.plugin.flutter.bridge.base.AdMethodHandler
import com.cleveradssolutions.plugin.flutter.sdk.AdContentInfoMethodHandler
import com.cleveradssolutions.plugin.flutter.sdk.OnAdImpressionListenerHandler
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturn
import com.cleveradssolutions.plugin.flutter.util.success
import com.cleveradssolutions.sdk.screen.CASAppOpen
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/app_open"

class AppOpenMethodHandler(
    binding: FlutterPlugin.FlutterPluginBinding,
    private val contextService: CASFlutterContext,
    contentInfoHandler: AdContentInfoMethodHandler
) : AdMethodHandler<CASAppOpen>(binding, CHANNEL_NAME, contentInfoHandler) {

    override fun initInstance(id: String): Ad<CASAppOpen> {
        val context = contextService.getContext()
        val appOpen = CASAppOpen(context, id)
        val contentInfoId = "app_open_$id"
        appOpen.contentCallback = ScreenAdContentCallbackHandler(this, id, contentInfoId)
        appOpen.onImpressionListener = OnAdImpressionListenerHandler(this, id, contentInfoId)
        return Ad(appOpen, id, contentInfoId)
    }

    override fun onMethodCall(
        instance: Ad<CASAppOpen>,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        when (call.method) {
            "isAutoloadEnabled" -> isAutoloadEnabled(instance.ad, result)
            "setAutoloadEnabled" -> setAutoloadEnabled(instance.ad, call, result)
            "isAutoshowEnabled" -> isAutoshowEnabled(instance.ad, result)
            "setAutoshowEnabled" -> setAutoshowEnabled(instance.ad, call, result)
            "isLoaded" -> isLoaded(instance.ad, result)
            "getContentInfo" -> getContentInfo(instance.contentInfoId, result)
            "load" -> load(instance.ad, result)
            "show" -> show(instance.ad, result)
            "destroy" -> destroy(instance, result)
            else -> super.onMethodCall(instance, call, result)
        }
    }

    private fun isAutoloadEnabled(appOpen: CASAppOpen, result: MethodChannel.Result) {
        result.success(appOpen.isAutoloadEnabled)
    }

    private fun setAutoloadEnabled(
        appOpen: CASAppOpen,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            appOpen.isAutoloadEnabled = it
        }
    }

    private fun isAutoshowEnabled(appOpen: CASAppOpen, result: MethodChannel.Result) {
        result.success(appOpen.isAutoshowEnabled)
    }

    private fun setAutoshowEnabled(
        appOpen: CASAppOpen,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            appOpen.isAutoshowEnabled = it
        }
    }

    private fun isLoaded(appOpen: CASAppOpen, result: MethodChannel.Result) {
        result.success(appOpen.isLoaded)
    }

    private fun getContentInfo(contentInfoId: String, result: MethodChannel.Result) {
        result.success(contentInfoId)
    }

    private fun load(appOpen: CASAppOpen, result: MethodChannel.Result) {
        appOpen.load(contextService.getContext())

        result.success()
    }

    private fun show(appOpen: CASAppOpen, result: MethodChannel.Result) {
        val activity = contextService.getActivity()

        appOpen.show(activity)

        result.success()
    }

    private fun destroy(ad: Ad<CASAppOpen>, result: MethodChannel.Result) {
        destroy(ad)
        ad.ad.destroy()

        result.success()
    }

}

File: ios/Classes/SDK/Screen/AppOpenMethodHandler.swift
//
//  AppOpenMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2025 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

private let channelName = "cleveradssolutions/app_open"

class AppOpenMethodHandler: AdMethodHandler<CASAppOpen> {
    private var contentDelegate: ScreenContentDelegateHandler?
    private var impressionDelegate: ImpressionDelegateHandler?
    
    init(with registrar: FlutterPluginRegistrar, _ contentInfoMethodHandler: AdContentInfoMethodHandler) {
        super.init(with: registrar, on: channelName, contentInfoMethodHandler)
    }

    override func initInstance(_ id: String) -> AdMethodHandler<CASAppOpen>.Ad {
        let appOpen = CASAppOpen(casID: id)
        let contentInfoId = "app_open_\(id)"
        let contentDelegate = ScreenContentDelegateHandler(self, id, contentInfoId)
        self.contentDelegate = contentDelegate
        appOpen.delegate = contentDelegate
        let impressionDelegate = ImpressionDelegateHandler(self, id, contentInfoId)
        appOpen.impressionDelegate = impressionDelegate
        self.impressionDelegate = impressionDelegate
        return Ad(ad: appOpen, id: id, contentInfoId: contentInfoId)
    }

    override func onMethodCall(_ instance: Ad, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "isAutoloadEnabled":
            isAutoloadEnabled(instance.ad, result)
        case "setAutoloadEnabled":
            setAutoloadEnabled(instance.ad, call, result)
        case "isAutoshowEnabled":
            isAutoshowEnabled(instance.ad, result)
        case "setAutoshowEnabled":
            setAutoshowEnabled(instance.ad, call, result)
        case "isLoaded":
            isLoaded(instance.ad, result)
        case "getContentInfo":
            getContentInfo(instance.contentInfoId, result)
        case "load":
            load(instance.ad, result)
        case "show":
            show(instance.ad, result)
        case "destroy":
            destroy(instance, result)
        default: super.onMethodCall(call, result)
        }
    }

    private func isAutoloadEnabled(_ appOpen: CASAppOpen, _ result: @escaping FlutterResult) {
        result(appOpen.isAutoloadEnabled)
    }

    private func setAutoloadEnabled(_ appOpen: CASAppOpen, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("isEnabled", result) { isEnabled in
            appOpen.isAutoloadEnabled = isEnabled
        }
    }

    private func isAutoshowEnabled(_ appOpen: CASAppOpen, _ result: @escaping FlutterResult) {
        result(appOpen.isAutoshowEnabled)
    }

    private func setAutoshowEnabled(_ appOpen: CASAppOpen, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("isEnabled", result) { isEnabled in
            appOpen.isAutoshowEnabled = isEnabled
        }
    }

    private func isLoaded(_ appOpen: CASAppOpen, _ result: @escaping FlutterResult) {
        result(appOpen.isAdLoaded)
    }

    private func getContentInfo(_ contentInfoId: String, _ result: @escaping FlutterResult) {
        result(contentInfoId)
    }

    private func load(_ appOpen: CASAppOpen, _ result: @escaping FlutterResult) {
        appOpen.loadAd()

        result(nil)
    }

    private func show(_ appOpen: CASAppOpen, _ result: @escaping FlutterResult) {
        appOpen.present()

        result(nil)
    }

    private func destroy(_ ad: AdMethodHandler<CASAppOpen>.Ad, _ result: @escaping FlutterResult) {
        contentDelegate = nil
        impressionDelegate = nil
        destroy(ad: ad)
        ad.ad.destroy()

        result(nil)
    }
}
